### PR TITLE
Handcuffs and aggro grab restrict you from reaching for comms

### PR DESCRIPTION
**Pull request by Tokorizo**
## About The Pull Request
Original PR: https://github.com/Skyrat-SS13/Skyrat-tg/pull/1285

Pretty simple balance change, if your hands are cuffed, or you are the victim of an aggressive grab you cannot reach for your headset to call for help.

## Why It's Good For The Game

A lot of RP is ruined mostly from a single ;HELP during combat. Not only does it cause future RP to be worst now to active sec offs patrolling, but the inevitable timer and alert it cause for anyone who manages to hear it. The purpose is to allow antags and even nonantags the ability to silence someone without having to murder them and possibly allow for RP that is slower-paced and developed.

## Changelog

:cl:Azarak
balance:  You can no longer use radio while cuffed or aggo-grabbed
/:cl:

### CODE REVIEW

**Comment by Kapu1178** [comment]
Can't this just be enforced by the rules? Handcuff logging exists.

**Comment by Tokorizo** [comment]
> Can't this just be enforced by the rules? Handcuff logging exists.

Not sure what you mean by this, like ahelping people for using the radio during combat? That's just extra work when we can just code a solution to it. We don't need to make staff log dive stuff we can just fix it with code.

File: code/modules/mob/living/carbon/carbon_say.dm [hunk 1]
```diff
@@ -23,3 +23,13 @@
 		return T.could_speak_language(language)
 	else
 		return initial(language.flags) & TONGUELESS_SPEECH
+
+/mob/living/carbon/radio(message, list/message_mods = list(), list/spans, language)
+	if((message_mods[MODE_HEADSET] || message_mods[RADIO_EXTENSION]))
```
Comment (SabreML) [inline code/modules/mob/living/carbon/carbon_say.dm:28]: No need for the extra brackets here:
```suggestion
	if(message_mods[MODE_HEADSET] || message_mods[RADIO_EXTENSION])
```

File: code/modules/mob/living/carbon/carbon_say.dm [hunk 2]
```diff
@@ -23,3 +23,13 @@
...
+		if(handcuffed)// If we're handcuffed, we can't press the button
+			to_chat(src, "<span class='warning'>You can't use the radio while handcuffed!</span>")
```
Comment (SabreML) [inline code/modules/mob/living/carbon/carbon_say.dm:30]:
```suggestion
			to_chat(src, span_warning("You can't use the radio while handcuffed!"))
```

File: code/modules/mob/living/carbon/carbon_say.dm [hunk 3]
```diff
@@ -23,3 +23,13 @@
...
+			return ITALICS | REDUCE_RANGE
+		if(pulledby && pulledby.grab_state)
+			to_chat(src, "<span class='warning'>You can't use the radio while aggressively grabbed!</span>")
```
Comment (SabreML) [inline code/modules/mob/living/carbon/carbon_say.dm:33]:
```suggestion
			to_chat(src, span_warning("You can't use the radio while aggressively grabbed!"))
```

File: code/modules/mob/living/carbon/carbon_say.dm [hunk 4]
```diff
@@ -23,3 +23,13 @@
 		return T.could_speak_language(language)
 	else
 		return initial(language.flags) & TONGUELESS_SPEECH
+
+/mob/living/carbon/radio(message, list/message_mods = list(), list/spans, language)
+	if((message_mods[MODE_HEADSET] || message_mods[RADIO_EXTENSION]))
+		if(handcuffed)// If we're handcuffed, we can't press the button
+			to_chat(src, "<span class='warning'>You can't use the radio while handcuffed!</span>")
+			return ITALICS | REDUCE_RANGE
+		if(pulledby && pulledby.grab_state)
```
Comment (SabreML) [inline code/modules/mob/living/carbon/carbon_say.dm:32]: `pulledby?.grab_state` acts the same as `pulledby && pulledby_grab_state`, and it's a bit neater:
```suggestion
		if(pulledby?.grab_state)
```

**Review by SabreML — COMMENTED** [review]
Just some styling things.

**Review by SabreML — COMMENTED** [review]
In order to make radio implants still work while you're restrained:
*(This won't work in a regular review format so I'm doing it like this instead)*
<hr>

<details>
<summary><b>code\modules\mob\living\carbon\carbon_say.dm</b></summary>

```diff
		return T.could_speak_language(language)
	else
		return initial(language.flags) & TONGUELESS_SPEECH

-/mob/living/carbon/radio(message, list/message_mods = list(), list/spans, language)
-	if(message_mods[MODE_HEADSET] || message_mods[RADIO_EXTENSION])
-		if(handcuffed)// If we're handcuffed, we can't press the button
-			to_chat(src, span_warning("You can't use the radio while handcuffed!"))
-			return ITALICS | REDUCE_RANGE
-		if(pulledby?.grab_state)
-			to_chat(src, span_warning("You can't use the radio while aggressively grabbed!"))
-			return ITALICS | REDUCE_RANGE
-	return ..()
```
(Or `git checkout upstream/master code\modules\mob\living\carbon\carbon_say.dm`)
</details>
<hr>

<details>
<summary><b>code\game\objects\items\devices\radio\radio.dm</b></summary>

```diff
	/// If true, broadcasts will be large and BOLD.
	var/use_command = FALSE
	/// If true, use_command can be toggled at will.
	var/command = FALSE

+	/// If true, the radio can be used even when the user is restrained.
+	var/handsfree = FALSE
+
	///makes anyone who is talking through this anonymous.
	var/anonymize = FALSE

	/// Encryption key handling
	var/obj/item/encryptionkey/keyslot
```
```diff
/obj/item/radio/talk_into(atom/movable/talking_movable, message, channel, list/spans, datum/language/language, list/message_mods)
+	if(iscarbon(talking_movable) && (message_mods[MODE_HEADSET] || message_mods[RADIO_EXTENSION]) && !handsfree)
+		var/mob/living/carbon/talking_carbon = talking_movable
+		if(talking_carbon.handcuffed)// If we're handcuffed, we can't press the button
+			to_chat(talking_carbon, span_warning("You can't use the radio while handcuffed!"))
+			return ITALICS | REDUCE_RANGE
+		if(talking_carbon.pulledby?.grab_state)
+			to_chat(talking_carbon, span_warning("You can't use the radio while aggressively grabbed!"))
+			return ITALICS | REDUCE_RANGE
+
	if(HAS_TRAIT(talking_movable, TRAIT_SIGN_LANG)) //Forces Sign Language users to wear the translation gloves to speak over radios
		var/mob/living/carbon/mute = talking_movable
```
</details>
<hr>

<details>
<summary><b>code\game\objects\items\implants\implant_misc.dm</b></summary>

```diff
/obj/item/implanter/emp
	name = "implanter (EMP)"
	imp_type = /obj/item/implant/emp

+/obj/item/radio/implanted
+	handsfree = TRUE
+
/obj/item/implant/radio
	name = "internal radio implant"
	activated = TRUE
-	var/obj/item/radio/radio
+	var/obj/item/radio/implanted/radio
	var/radio_key
	var/subspace_transmission = FALSE
	icon = 'icons/obj/radio.dmi'
	icon_state = "walkietalkie"

/obj/item/implant/radio/activate()
	. = ..()
	// needs to be GLOB.deep_inventory_state otherwise it won't open
	radio.ui_interact(usr, state = GLOB.deep_inventory_state)
```
</details>

**Review by SabreML — APPROVED** [review]

**Pull request merged**